File: 3561-remove-methods-from-project/remove-methods-from-project.cpp
class Solution {
public:
    void dfs(unordered_map<int, vector<int>> &g, vector<bool> &vis, int source) {

        vis[source] = true;

        for (int neighbour : g[source]) {

            if (!vis[neighbour]) {
                dfs(g, vis, neighbour);
            }
        }
    }

    vector<int> remainingMethods(int n, int k, vector<vector<int>>& in) {

        unordered_map<int, vector<int>> g;
        unordered_map<int, vector<int>> p;

        for (auto a : in) {
            int s = a[0];
            int d = a[1];

            g[s].push_back(d);
            p[d].push_back(s);
        }

        vector<bool> vis(n, false);
        vector<int> ans;

        dfs(g, vis, k);

        bool flag = false;

        for (int i = 0; i < n; i++) {

            if (vis[i]) {

                vector<int> parents = p[i];

                for (int par : parents) {

                    if (!vis[par]) {
                        flag = true;
                        break;
                    }
                }

                if (flag) break;
            }
        }

        if (flag) {
            for (int i = 0; i < n; i++)
                ans.push_back(i);
        }
        else {
            for (int i = 0; i < n; i++) {
                if (!vis[i])
                    ans.push_back(i);
            }
        }

        return ans;
    }
};

File: 3561-remove-methods-from-project/remove-methods-from-project.py
from collections import defaultdict


class Solution:

    def dfs(self, g, vis, source):

        vis[source] = True

        for neighbour in g[source]:

            if not vis[neighbour]:
                self.dfs(g, vis, neighbour)

    def remainingMethods(self, n, k, invocations):

        g = defaultdict(list)
        p = defaultdict(list)

        for a in invocations:

            s = a[0]
            d = a[1]

            g[s].append(d)
            p[d].append(s)

        vis = [False] * n
        ans = []

        self.dfs(g, vis, k)

        flag = False

        for i in range(n):

            if vis[i]:

                parents = p[i]

                for par in parents:

                    if not vis[par]:
                        flag = True
                        break

                if flag:
                    break

        if flag:

            for i in range(n):
                ans.append(i)

        else:

            for i in range(n):

                if not vis[i]:
                    ans.append(i)

        return ans